quickwit-indexing/src/actors/merge_planner.rs

```diff
@@ -66,46 +70,88 @@ impl Handler<NewSplits> for MergePlanner {
         message: NewSplits,
         ctx: &ActorContext<Self>,
     ) -> Result<(), ActorExitStatus> {
-        let mut has_new_young_split = false;
-        for split in message.new_splits {
-            if self.merge_policy.is_mature(&split) {
-                info!(split_id=%split.split_id(), num_docs=split.num_docs, size_in_bytes=split.uncompressed_docs_size_in_bytes, "mature-split");
-                continue;
-            }
-            self.young_splits.push(split);
-            has_new_young_split = true;
-        }
-        if has_new_young_split {
-            self.send_operations(ctx).await?;
+        let mut target_partition_ids = Vec::new();
+
+        let partitioned_new_young_splits = message
+            .new_splits
+            .into_iter()
+            .filter(|split| {
+                !self.merge_policy.is_mature(split) || {
+                    info!(
+                        split_id=%split.split_id(),
+                        num_docs=split.num_docs,
+                        num_bytes=split.uncompressed_docs_size_in_bytes,
+                        "Split is mature."
+                    );
+                    false
+                }
+            })
+            .group_by(|split| split.partition_id);
+
+        for (partition_id, new_young_splits) in &partitioned_new_young_splits {
+            let young_splits = self
+                .partitioned_young_splits
+                .entry(partition_id)
+                .or_default();
+            young_splits.extend(new_young_splits);
+            target_partition_ids.push(partition_id);
         }
+        self.send_merge_ops(ctx, &target_partition_ids).await?;
         Ok(())
     }
 }
 
 impl MergePlanner {
     pub fn new(
-        young_splits: Vec<SplitMetadata>,
+        pipeline_id: IndexingPipelineId,
+        published_splits: Vec<SplitMetadata>,
         merge_policy: Arc<dyn MergePolicy>,
         merge_split_downloader_mailbox: Mailbox<MergeSplitDownloader>,
     ) -> MergePlanner {
+        let mut partitioned_young_splits = HashMap::new();
+        for (partition_id, young_splits) in &published_splits
+            .into_iter()
+            .filter(|split| {
+                belongs_to_pipeline(&pipeline_id, split) && !merge_policy.is_mature(split)
+            })
+            .group_by(|split| split.partition_id)
+        {
+            partitioned_young_splits.insert(partition_id, young_splits.collect());
```

Comment:
A for loop is sometimes more readable than chaining, I think it's ok here.

but fyi, we can do a collect directly here.

```rust
    let data = vec![1, 3, 2, 2, 1, 0, 1, 2, 1, 2, 3, 4];
    let groups: HashMap<usize, Vec<usize>>  = data.iter()
        .group_by(|element| **element)
        .into_iter()
        .map(|(key, group)| (key, group.cloned().collect()))
        .collect();
```

Reply:
I'm trying to extend the already existing vectors of young splits without allocating a temporary vector for each group. `Vec::extends` accepts an iterator, and I'm taking advantage of this API here.